MDEV-28730 Remove internal parser usage from InnoDB fts

Thirunarayanan opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-28730*

## Description
Remove internal parser/SQL-graph usage and migrate FTS paths to QueryExecutor

Introduced QueryExecutor (row0query.{h,cc}) and FTSQueryExecutor abstractions for
clustered, secondary scans and DML.

Refactored fetch/optimize code to use QueryExecutor::read(), read_by_index()
with RecordCallback, replacing SQL graph flows

Added CommonTableReader and ConfigReader callbacks for common/CONFIG tables

Implemented fts_index_fetch_nodes(trx, index, word, user_arg, FTSRecordProcessor, compare_mode)
and rewrote fts_optimize_write_word() to delete/insert via executor with fts_aux_data_t

Removed fts_doc_fetch_by_doc_id() and FTS_FETCH_DOC_BY_ID_* macros, updating callers to
fts_query_fetch_document()

Tightened fts_select_index{,_by_range,by_hash} return type to uint8_t;
Removed fts0sql.cc and eliminated fts_table_t from fts_query_t/fts_optimize_t.*


## Release Notes
Removed the sql parser usage from fulltext subsystem

## How can this PR be tested?
For QA purpose, Run RQG testing involving Fulltext subsystem

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4443) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=4443) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=4443) it.</sub>

grooverdan commented:
In addition to the CI failures needing correcting, does this mean `storage/innobase/pars/` and the `fts_parse_sql` at least can be removed from `storage/innobase/include/fts0priv.h`?

Great to see the parser going away.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,634 @@
+  ut_ad(!dict_index_is_clust(sec_index));
+
+  dict_index_t *clust_index= dict_table_get_first_index(table);
+  if (!clust_index) return DB_ERROR;
```
The function should be declared with `__attribute__((nonnull))` and the related assertions removed. Why is `search_tuple` not a pointer to `const`? There is a space before each unary `*` but after the unary `&`. That is inconsistent.

Do we really need the `if` condition here? There should be no code path that would add a table to `dict_sys` such that it does not have any indexes.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,634 @@
+  /* Extract primary key from secondary index record */
+  dtuple_t *clust_tuple= row_build_row_ref(ROW_COPY_DATA, sec_index,
+                                           sec_rec, m_heap);
+  if (!clust_tuple) return false;
```
This looks like dead code. If `ref = dtuple_create(heap, ref_len);` returned `nullptr` to `row_build_row_ref()`, the subsequent statement `dict_index_copy_types(ref, clust_index, ref_len);` should crash. So, a `nullptr` return value should be impossible and this should be dead code.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,634 @@
+    }
+  }
+  clust_mtr.commit();
+  return record_processed ? continue_processing : true;
```
I find it a form of obfuscation to use the ternary operator when any of the parameters is `true` or `false´. This is equivalent to the following:
```c++
  return !record_processed || continue_processing;
```

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,634 @@
+  bool record_processed= false;
+  dberr_t clust_err= btr_pcur_open(clust_tuple, PAGE_CUR_LE,
+                                   BTR_SEARCH_LEAF, &clust_pcur,
+                                   &clust_mtr);
```
We are discarding any error code here. If I understood it correctly, we would return `true` if an error occurs. If the function comment `@return true if should continue processing, false to break` is valid, this could lead to an infinite loop on corruption.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,634 @@
+{
+  dict_index_t *index= dict_table_get_first_index(table);
+  btr_pcur_t pcur;
+  mtr_t mtr(m_trx);
```
Why is `mtr` not a  data member of `QueryExecutor`?

dr-m commented on storage/innobase/include/fts0exec.h:
```diff
@@ -0,0 +1,306 @@
+*******************************************************/
+
+#ifndef INNOBASE_FTS0QUERY_H
+#define INNOBASE_FTS0QUERY_H
```
Let’s just use `#pragma once` like we do in many other files.

dr-m reviewed (commented):
Here are some quick initial comments.

dr-m commented on storage/innobase/handler/i_s.cc:
```diff
@@ -2660,100 +2657,62 @@ struct st_maria_plugin	i_s_innodb_ft_index_cache =
+  }
+  trx->free();
+  if (total_memory >= fts_result_cache_limit)
+    error= DB_FTS_EXCEED_RESULT_CACHE_LIMIT;
```
Shouldn’t this condition be checked inside the loop? Shouldn’t an `executor` member function return this error?

dr-m reviewed (commented):
Here are some more comments. The error propagation is better now, but I would like to see some more effort to avoid the number of `dict_sys.latch` acquisitions. This should be tested as well, in a custom benchmark.

Even though we are adding quite a bit of code, I was pleasantly surprised that the size of a x86-64 `CMAKE_BUILD_TYPE=RelWithDebInfo` executable would increase by only 20 KiB. I believe that removing the InnoDB SQL parser (once some more code has been refactored) would remove more code than that.

Thirunarayanan commented on storage/innobase/include/fts0exec.h:
```diff
@@ -0,0 +1,294 @@
+  QueryExecutor *m_executor;
+  bool m_dict_locked;
+  const dict_index_t *m_fts_index;
+  const dict_table_t *m_fts_table;
```
There can be multiple fulltext index can exist for the table. So I removed fts_index from constructor.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,705 @@
+  mem_heap_t* version_heap= nullptr;
+  mem_heap_t* offsets_heap= nullptr;
+  rec_offs* offsets= nullptr;
+  rec_offs* version_offsets= nullptr;
```
Why are we not allocating `offsets` and `version_offsets` from the stack, and forcing memory to be allocated from the heap every single time?

Could we cope with a single `mem_heap_t` object here? Or none at all? This function is supposed to be run on some tables whose schema we know in advance, right? Can this ever be run on a user-defined table? Unfortunately, in none of the callers of `QueryExecutor::process_record_with_mvcc()` did I find any assertion on the table names or the schema.

dr-m commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,705 @@
+  if (rec_trx_id && !m_mtr.trx->read_view.changes_visible(rec_trx_id))
+  {
+    /* Compute clustered index offsets */
+    rec_t *version_rec= const_cast<rec_t*>(rec);
```
Are the `const_cast` really necessary? Is the variable `version_rec` necessary? Couldn’t we just assign directly to `result_rec` instead?

dr-m reviewed (commented):
I tried to review the MVCC logic, but I did not fully understand it.

It would be much more convenient to review this if the tables or indexes on which the new code is expected to be invoked were prominently documented in debug assertions.

dr-m commented on storage/innobase/row/row0sel.cc:
There already are very many parameters passed to this member function call. Adding 4 more parameters will complicate the call setup; I don’t think that any ABI would support passing this number of parameters in registers. Can we move most of the parameters to data members?

Do we need both `trx` and `thr`? Is `thr->graph->trx` not always accessible?

Instead of changing this interface, could we initialize a `row_prebuilt_t` in the code path that is missing it?

dr-m reviewed (commented):
The `QueryExecutor` bears some similarity with functions that operate on `row_prebuilt_t`. Could we unify them better? Would it be feasible to construct `row_prebuit_t` objects in the `FULLTEXT` subsystem, so that even higher-level functions such as `row_search_mvcc()` can be invoked?

mariadb-SaahilAlam commented on storage/innobase/fts/fts0opt.cc:
```diff
@@ -1680,175 +1277,129 @@ fts_optimize_free(
+    if (error == DB_SUCCESS)
+    {
+      /* There must be some nodes to read. */
+      ut_a(ib_vector_size(optim->words) > 0);
```
Hitting this Assertion
`# 2026-05-14T14:04:17 [56807] | InnoDB: Failing assertion: ib_vector_size(optim->words) > 0`
Bug raised:https://jira.mariadb.org/browse/MDEV-39616

Thirunarayanan replied:
Closed

mariadb-SaahilAlam commented on storage/innobase/page/page0cur.cc:
Found this Assertion:-
`# 2026-05-14T14:34:41 [1226396] | mariadbd: /data/Server/MDEV-28730_new/storage/innobase/page/page0cur.cc:87: int cmp_dtuple_rec_bytes(const rec_t*, const dict_index_t&, const dtuple_t&, int*, ulint): Assertion `page_rec_is_leaf(rec)' failed.`
Bug raised:https://jira.mariadb.org/browse/MDEV-39617

Thirunarayanan replied:
Closed

mariadb-SaahilAlam commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,445 @@
+                                     ULINT_UNDEFINED, &m_heap);
+
+  dberr_t err= DB_SUCCESS;
+  ulint cmpl_info= UPD_NODE_NO_ORD_CHANGE | UPD_NODE_NO_SIZE_CHANGE;
```
Crash found during testing 
```
Thread 3 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 3906580.3917832]
thd_storage_lock_wait (thd=0x0, value=11) at /data/Server/MDEV-28730_new/sql/sql_class.cc:451
451       thd->utime_after_lock+= value;
(rr) bt
#0  thd_storage_lock_wait (thd=0x0, value=11) at /data/Server/MDEV-28730_new/sql/sql_class.cc:451
#1  0x000056d4586cb8e3 in lock_sys_t::wait_resume (this=this@entry=0x56d459ee5c80 <lock_sys>, thd=0x0, start=start@entry=..., now=...) at /data/Server/MDEV-28730_new/storage/innobase/lock/lock0lock.cc:2108
#2  0x000056d4586c9128 in lock_wait (thr=thr@entry=0x1276380563c8) at /data/Server/MDEV-28730_new/storage/innobase/lock/lock0lock.cc:2410
#3  0x000056d45879c93d in row_mysql_handle_errors (new_err=new_err@entry=0x799151a0c4a0, trx=trx@entry=0x55aa7560ff00, thr=thr@entry=0x1276380563c8, savept=savept@entry=0x0) at /data/Server/MDEV-28730_new/storage/innobase/row/row0mysql.cc:698
#4  0x000056d4587d308d in row_search_mvcc<InnoDBPolicy<false, false> > (buf=buf@entry=0x799151a0cd90 "H\203>Y\324V", mode=mode@entry=PAGE_CUR_GE, prebuilt=0x127638056468, match_mode=match_mode@entry=0, direction=direction@entry=0)
    at /data/Server/MDEV-28730_new/storage/innobase/row/row0sel.cc:6051
#5  0x000056d45893409c in row_search_mvcc_callback_dml (direction=0, match_mode=0, prebuilt=<optimized out>, mode=PAGE_CUR_GE, callback=0x799151a0cd90) at /data/Server/MDEV-28730_new/storage/innobase/include/row0sel.h:202
#6  QueryExecutor::select_for_update (this=this@entry=0x799151a0ce70, table=table@entry=0x12763803fbf8, search_tuple=search_tuple@entry=0x799151a0cd20, callback=callback@entry=0x799151a0cd90)
    at /data/Server/MDEV-28730_new/storage/innobase/row/row0query.cc:331
#7  0x000056d458906f5f in FTSQueryExecutor::read_config_with_lock (this=this@entry=0x799151a0ce70, key=key@entry=0x56d458cb4922 "synced_doc_id", callback=...) at /data/Server/MDEV-28730_new/storage/innobase/fts/fts0exec.cc:535
#8  0x000056d4589044bf in fts_config_get_value (executor=executor@entry=0x799151a0ce70, table=table@entry=0x127638103a68, name=name@entry=0x56d458cb4922 "synced_doc_id", value=value@entry=0x799151a0ce50)
    at /data/Server/MDEV-28730_new/storage/innobase/fts/fts0config.cc:46
#9  0x000056d4586a3eb3 in i_s_fts_config_fill (thd=0x127638000d58, tables=<optimized out>) at /data/Server/MDEV-28730_new/storage/innobase/handler/i_s.cc:3161
#10 0x000056d45813ed14 in get_schema_tables_result (join=join@entry=0x1276380178a8, executed_place=executed_place@entry=PROCESSED_BY_JOIN_EXEC) at /data/Server/MDEV-28730_new/sql/sql_show.cc:9924
#11 0x000056d4581286d1 in JOIN::exec_inner (this=this@entry=0x1276380178a8) at /data/Server/MDEV-28730_new/sql/sql_select.cc:5086
#12 0x000056d4581288f1 in JOIN::exec (this=this@entry=0x1276380178a8) at /data/Server/MDEV-28730_new/sql/sql_select.cc:4913
#13 0x000056d4581278df in mysql_select (thd=thd@entry=0x127638000d58, tables=0x127638016628, fields=..., conds=0x0, og_num=0, order=<optimized out>, group=0x0, having=0x0, proc_param=0x0, select_options=2701396736, result=0x127638017880, 
    unit=0x127638005278, select_lex=0x127638015ea8) at /data/Server/MDEV-28730_new/sql/sql_select.cc:5439
#14 0x000056d458127ab4 in handle_select (thd=thd@entry=0x127638000d58, lex=lex@entry=0x127638005198, result=result@entry=0x127638017880, setup_tables_done_option=setup_tables_done_option@entry=0)
    at /data/Server/MDEV-28730_new/sql/sql_select.cc:636
#15 0x000056d45809d35e in execute_sqlcom_select (thd=thd@entry=0x127638000d58, all_tables=0x127638016628) at /data/Server/MDEV-28730_new/sql/sql_parse.cc:6214
#16 0x000056d4580a6dd3 in mysql_execute_command (thd=thd@entry=0x127638000d58, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=false) at /data/Server/MDEV-28730_new/sql/sql_parse.cc:3988
#17 0x000056d4580ac979 in mysql_parse (thd=thd@entry=0x127638000d58, rawbuf=<optimized out>, length=<optimized out>, parser_state=parser_state@entry=0x799151a0e3c0) at /data/Server/MDEV-28730_new/sql/sql_parse.cc:7942
#18 0x000056d4580adfcd in dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=0x127638000d58, packet=packet@entry=0x12763800b669 "\nSELECT COUNT(*) FROM INFORMATION_SCHEMA.INNODB_FT_CONFIG /* E_R Thread6 QNO 79 CON_ID 21 */ ", 
    packet_length=packet_length@entry=93, blocking=blocking@entry=true) at /data/Server/MDEV-28730_new/sql/sql_parse.cc:1898
#19 0x000056d4580af5d2 in do_command (thd=thd@entry=0x127638000d58, blocking=blocking@entry=true) at /data/Server/MDEV-28730_new/sql/sql_parse.cc:1432
#20 0x000056d45820859d in do_handle_one_connection (connect=<optimized out>, connect@entry=0x56d45b7f35b8, put_in_cache=put_in_cache@entry=true) at /data/Server/MDEV-28730_new/sql/sql_connect.cc:1503
#21 0x000056d4582087ba in handle_one_connection (arg=0x56d45b7f35b8) at /data/Server/MDEV-28730_new/sql/sql_connect.cc:1415
#22 0x0000764474ab2a94 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:447
#23 0x0000764474b3fa34 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```
Bug raised:-https://jira.mariadb.org/browse/MDEV-39621

Thirunarayanan replied:
This should be fixed in the latest patch.

mariadb-SaahilAlam commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -4192,7 +3665,7 @@ fts_sync(
 end_sync:
 	if (error == DB_SUCCESS && !sync->interrupted) {
-		error = fts_sync_commit(sync);
+		error = fts_sync_commit(&executor, sync);
```
hitting this assertion on  `origin/main-MDEV-28730 a9665958948732b39b5a71f2b66672a97883837b`

```
# 2026-05-15T16:50:34 [1679424] | 2026-05-15 16:50:01 22 [ERROR] InnoDB: ( Lock wait ) writing word node to FTS auxiliary index table test/table300_innodb_int_autoinc
# 2026-05-15T16:50:34 [1679424] | mariadbd: /data/Server/MDEV-28730_1/storage/innobase/include/trx0trx.h:1124: void trx_t::assert_freed() const: Assertion `!lock.wait_lock' failed.
```
RR trace is present on SDP :-
`/data/results/1778871863/MDEV-28730_05`

mariadb-SaahilAlam replied:
This is not reproducible on latest commit

mariadb-SaahilAlam commented on storage/innobase/fts/fts0exec.cc:
```diff
@@ -0,0 +1,886 @@
+{
+  dict_table_t *table= m_aux_tables[aux_index];
+  if (table == nullptr) return DB_TABLE_NOT_FOUND;
+  dberr_t err= m_executor.lock_table(table, mode);
```
Assertion found on origin/main-MDEV-28730 29c838cc95b532d26f91e327dc684c8973b218f1

```
`# 2026-05-18T12:58:16 [3307300] | mariadbd: /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:985: void trx_start_low(trx_t*, bool): Assertion `!read_write' failed.`

# 2026-05-18T12:59:09 [3307300] | #0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
# 2026-05-18T12:59:09 [3307300] | #1  __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
# 2026-05-18T12:59:09 [3307300] | #2  __GI___pthread_kill (threadid=<optimized out>, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
# 2026-05-18T12:59:09 [3307300] | #3  0x00007deb72e4526e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
# 2026-05-18T12:59:09 [3307300] | #4  0x00007deb72e288ff in __GI_abort () at ./stdlib/abort.c:79
# 2026-05-18T12:59:09 [3307300] | #5  0x00007deb72e2881b in __assert_fail_base (fmt=0x7deb72fd01e8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x6464bf127506 "!read_write", file=file@entry=0x6464bf07ab10 "/data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc", line=line@entry=985, function=function@entry=0x6464bf07ad38 "void trx_start_low(trx_t*, bool)") at ./assert/assert.c:94
# 2026-05-18T12:59:09 [3307300] | #6  0x00007deb72e3b507 in __assert_fail (assertion=0x6464bf127506 "!read_write", file=0x6464bf07ab10 "/data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc", line=985, function=0x6464bf07ad38 "void trx_start_low(trx_t*, bool)") at ./assert/assert.c:103
# 2026-05-18T12:59:09 [3307300] | #7  0x00006464bec810a7 in trx_start_low (trx=0x7deb70402e80, read_write=read_write@entry=true) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:985
# 2026-05-18T12:59:09 [3307300] | #8  0x00006464bec81274 in trx_start_if_not_started_low (trx=<optimized out>, read_write=read_write@entry=true) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:2216
# 2026-05-18T12:59:09 [3307300] | #9  0x00006464bed9cb2c in QueryExecutor::lock_table (this=this@entry=0x7deb7015b190, table=0x75eb38496c58, mode=mode@entry=LOCK_IS) at /data/Server/MDEV-28730_02/storage/innobase/row/row0query.cc:154
# 2026-05-18T12:59:09 [3307300] | #10 0x00006464bed6ddd8 in FTSQueryExecutor::lock_aux_tables (this=this@entry=0x7deb7015b190, aux_index=aux_index@entry=0 '\000', mode=mode@entry=LOCK_IS) at /data/Server/MDEV-28730_02/storage/innobase/fts/fts0exec.cc:161
# 2026-05-18T12:59:09 [3307300] | #11 0x00006464bed6fca4 in FTSQueryExecutor::read_aux (this=this@entry=0x7deb7015b190, aux_index=aux_index@entry=0 '\000', start_word=start_word@entry=0x7deb7015b100, callback=...) at /data/Server/MDEV-28730_02/storage/innobase/fts/fts0exec.cc:863
# 2026-05-18T12:59:09 [3307300] | #12 0x00006464beb0919a in i_s_fts_index_table_fill_selected (executor=executor@entry=0x7deb7015b190, reader=..., selected=selected@entry=0 '\000', word=word@entry=0x7deb7015b100) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:2679
# 2026-05-18T12:59:09 [3307300] | #13 0x00006464beb0ca39 in i_s_fts_read_aux_index_words (executor=executor@entry=0x7deb7015b190, reader=..., selected=selected@entry=0 '\000', thd=thd@entry=0x75eb4400b358, tables=tables@entry=0x75eb44020c18, conv_str=conv_str@entry=0x7deb7015b3a0, heap=0x75eb4404f8a0, words=0x75eb4404f938) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:2851
# 2026-05-18T12:59:09 [3307300] | #14 0x00006464beb0e1b9 in i_s_fts_index_table_fill_one_index (index=index@entry=0x75eb3828b638, thd=thd@entry=0x75eb4400b358, conv_str=conv_str@entry=0x7deb7015b3a0, tables=tables@entry=0x75eb44020c18) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:2942
# 2026-05-18T12:59:09 [3307300] | #15 0x00006464beb0e34e in i_s_fts_index_table_fill (thd=0x75eb4400b358, tables=0x75eb44020c18) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:3000
# 2026-05-18T12:59:09 [3307300] | #16 0x00006464be5a6d14 in get_schema_tables_result (join=join@entry=0x75eb440220b0, executed_place=executed_place@entry=PROCESSED_BY_JOIN_EXEC) at /data/Server/MDEV-28730_02/sql/sql_show.cc:9924
# 2026-05-18T12:59:09 [3307300] | #17 0x00006464be5906d1 in JOIN::exec_inner (this=this@entry=0x75eb440220b0) at /data/Server/MDEV-28730_02/sql/sql_select.cc:5086
# 2026-05-18T12:59:09 [3307300] | #18 0x00006464be5908f1 in JOIN::exec (this=this@entry=0x75eb440220b0) at /data/Server/MDEV-28730_02/sql/sql_select.cc:4913
# 2026-05-18T12:59:09 [3307300] | #19 0x00006464be58f8df in mysql_select (thd=thd@entry=0x75eb4400b358, tables=0x75eb44020c18, fields=..., conds=0x0, og_num=0, order=<optimized out>, group=0x0, having=0x0, proc_param=0x0, select_options=2701396736, result=0x75eb44022088, unit=0x75eb4400f878, select_lex=0x75eb44020498) at /data/Server/MDEV-28730_02/sql/sql_select.cc:5439
# 2026-05-18T12:59:09 [3307300] | #20 0x00006464be58fab4 in handle_select (thd=thd@entry=0x75eb4400b358, lex=lex@entry=0x75eb4400f798, result=result@entry=0x75eb44022088, setup_tables_done_option=setup_tables_done_option@entry=0) at /data/Server/MDEV-28730_02/sql/sql_select.cc:636
# 2026-05-18T12:59:09 [3307300] | #21 0x00006464be50535e in execute_sqlcom_select (thd=thd@entry=0x75eb4400b358, all_tables=0x75eb44020c18) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:6214
# 2026-05-18T12:59:09 [3307300] | #22 0x00006464be50edd3 in mysql_execute_command (thd=thd@entry=0x75eb4400b358, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=false) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:3988
# 2026-05-18T12:59:09 [3307300] | #23 0x00006464be514979 in mysql_parse (thd=thd@entry=0x75eb4400b358, rawbuf=<optimized out>, length=<optimized out>, parser_state=parser_state@entry=0x7deb7015c3c0) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:7942
# 2026-05-18T12:59:09 [3307300] | #24 0x00006464be515fcd in dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=0x75eb4400b358, packet=packet@entry=0x75eb44015c39 "\nSELECT COUNT(*) FROM INFORMATION_SCHEMA.INNODB_FT_INDEX_TABLE\nLIMIT 500 /* E_R Thread2 QNO 23 CON_ID 19 */ ", packet_length=packet_length@entry=108, blocking=blocking@entry=true) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:1898
# 2026-05-18T12:59:09 [3307300] | #25 0x00006464be5175d2 in do_command (thd=thd@entry=0x75eb4400b358, blocking=blocking@entry=true) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:1432
# 2026-05-18T12:59:09 [3307300] | #26 0x00006464be67059d in do_handle_one_connection (connect=<optimized out>, connect@entry=0x6464c2354d68, put_in_cache=put_in_cache@entry=true) at /data/Server/MDEV-28730_02/sql/sql_connect.cc:1503
# 2026-05-18T12:59:09 [3307300] | #27 0x00006464be6707ba in handle_one_connection (arg=0x6464c2354d68) at /data/Server/MDEV-28730_02/sql/sql_connect.cc:1415
# 2026-05-18T12:59:09 [3307300] | #28 0x00007deb72e9ca94 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:447
# 2026-05-18T12:59:09 [3307300] | #29 0x00007deb72f29a34 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```
RRtrace is present on SDP :-
`/data/results/1779134248/MDEV-28730_06
`

mariadb-SaahilAlam replied:
this is not reproducible on latest commit

mariadb-SaahilAlam commented on storage/innobase/handler/i_s.cc:
```diff
@@ -2890,6 +2817,90 @@ i_s_fts_index_table_fill_one_fetch(
+      if (UNIV_LIKELY(error == DB_SUCCESS ||
+                      error == DB_FTS_EXCEED_RESULT_CACHE_LIMIT))
+      {
+        fts_sql_commit(trx);
```
Assertion found on release build, please check if its related:-
`origin/main-MDEV-28730 29c838cc95b532d26f91e327dc684c8973b218f1`
```
# 2026-05-19T01:01:00 [2370519] | 2026-05-19 01:00:56 0x71385c1ff6c0  InnoDB: Assertion failure in file /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc line 1427
# 2026-05-19T01:01:00 [2370519] | InnoDB: Failing assertion: UT_LIST_GET_LEN(lock.trx_locks) == 0
```
 
```
# 2026-05-19T01:01:40 [2370519] | #0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
# 2026-05-19T01:01:40 [2370519] | #1  __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
# 2026-05-19T01:01:40 [2370519] | #2  __GI___pthread_kill (threadid=<optimized out>, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
# 2026-05-19T01:01:40 [2370519] | #3  0x000071385ee4526e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
# 2026-05-19T01:01:40 [2370519] | #4  0x000071385ee288ff in __GI_abort () at ./stdlib/abort.c:79
# 2026-05-19T01:01:40 [2370519] | #5  0x00005b1ae6645769 in ut_dbg_assertion_failed (expr=expr@entry=0x5b1ae7208c70 "UT_LIST_GET_LEN(lock.trx_locks) == 0", file=file@entry=0x5b1ae7208ac0 "/data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc", line=line@entry=1427) at /data/Server/MDEV-28730_02/storage/innobase/ut/ut0dbg.cc:60
# 2026-05-19T01:01:40 [2370519] | #6  0x00005b1ae6644a9f in trx_t::commit_in_memory (mtr=0x71385c1fcbf0, this=0x71385c202180) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:1427
# 2026-05-19T01:01:40 [2370519] | #7  trx_t::commit_persist (this=this@entry=0x71385c202180) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:1622
# 2026-05-19T01:01:40 [2370519] | #8  0x00005b1ae6e5ce7f in trx_t::commit (this=0x71385c202180) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:1631
# 2026-05-19T01:01:40 [2370519] | #9  trx_commit_for_mysql (trx=trx@entry=0x71385c202180) at /data/Server/MDEV-28730_02/storage/innobase/trx/trx0trx.cc:1746
# 2026-05-19T01:01:40 [2370519] | #10 0x00005b1ae6db264d in i_s_fts_read_aux_index_words (words=0x69382c03f8d0, heap=<optimized out>, conv_str=<optimized out>, tables=<optimized out>, thd=<optimized out>, selected=0 '\000', reader=..., executor=0x71385c1fd0b0) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:2857
# 2026-05-19T01:01:40 [2370519] | #11 i_s_fts_index_table_fill_one_index (index=index@entry=0x6938201381e0, thd=thd@entry=0x69382c000fd8, conv_str=conv_str@entry=0x71385c1fd310, tables=tables@entry=0x69382c014198) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:2942
# 2026-05-19T01:01:40 [2370519] | #12 0x00005b1ae6db2d64 in i_s_fts_index_table_fill (thd=0x69382c000fd8, tables=0x69382c014198) at /data/Server/MDEV-28730_02/storage/innobase/handler/i_s.cc:3000
# 2026-05-19T01:01:40 [2370519] | #13 0x00005b1ae6910421 in get_schema_tables_result (join=join@entry=0x69382c016010, executed_place=executed_place@entry=PROCESSED_BY_JOIN_EXEC) at /data/Server/MDEV-28730_02/sql/sql_show.cc:9924
# 2026-05-19T01:01:40 [2370519] | #14 0x00005b1ae68f120d in JOIN::exec_inner (this=this@entry=0x69382c016010) at /data/Server/MDEV-28730_02/sql/sql_select.cc:5086
# 2026-05-19T01:01:40 [2370519] | #15 0x00005b1ae68f1be2 in JOIN::exec (this=this@entry=0x69382c016010) at /data/Server/MDEV-28730_02/sql/sql_select.cc:4913
# 2026-05-19T01:01:40 [2370519] | #16 0x00005b1ae68f0155 in mysql_select (thd=thd@entry=0x69382c000fd8, tables=0x69382c014198, fields=..., conds=0x69382c014f28, og_num=1, order=<optimized out>, group=0x0, having=0x0, proc_param=0x0, select_options=2701396736, result=0x69382c015fe8, unit=0x69382c005320, select_lex=0x69382c013808) at /data/Server/MDEV-28730_02/sql/sql_select.cc:5439
# 2026-05-19T01:01:40 [2370519] | #17 0x00005b1ae68f0837 in handle_select (thd=thd@entry=0x69382c000fd8, lex=lex@entry=0x69382c005240, result=result@entry=0x69382c015fe8, setup_tables_done_option=setup_tables_done_option@entry=0) at /data/Server/MDEV-28730_02/sql/sql_select.cc:636
# 2026-05-19T01:01:40 [2370519] | #18 0x00005b1ae6866f34 in execute_sqlcom_select (thd=thd@entry=0x69382c000fd8, all_tables=0x69382c014198) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:6214
# 2026-05-19T01:01:40 [2370519] | #19 0x00005b1ae68749d6 in mysql_execute_command (thd=thd@entry=0x69382c000fd8, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=false) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:3988
# 2026-05-19T01:01:40 [2370519] | #20 0x00005b1ae6876083 in mysql_parse (thd=0x69382c000fd8, rawbuf=<optimized out>, length=<optimized out>, parser_state=<optimized out>) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:7942
# 2026-05-19T01:01:40 [2370519] | #21 0x00005b1ae6878362 in dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=0x69382c000fd8, packet=packet@entry=0x69382c008f09 "\nSELECT word, doc_count, doc_id\nFROM INFORMATION_SCHEMA.INNODB_FT_INDEX_TABLE\nWHERE word LIKE CONCAT( SUBSTRING( 'tell', 1, 2 ), '%' )\nORDER BY word LIMIT 100 /* E_R Thread2 QNO 9 CON_ID 19 */ ", packet_length=packet_length@entry=193, blocking=blocking@entry=true) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:1997
# 2026-05-19T01:01:40 [2370519] | #22 0x00005b1ae6879a96 in do_command (thd=thd@entry=0x69382c000fd8, blocking=blocking@entry=true) at /data/Server/MDEV-28730_02/sql/sql_parse.cc:1432
# 2026-05-19T01:01:40 [2370519] | #23 0x00005b1ae699f08d in do_handle_one_connection (connect=<optimized out>, put_in_cache=true) at /data/Server/MDEV-28730_02/sql/sql_connect.cc:1503
# 2026-05-19T01:01:40 [2370519] | #24 0x00005b1ae699f385 in handle_one_connection (arg=0x5b1ae8ad1d48) at /data/Server/MDEV-28730_02/sql/sql_connect.cc:1415
# 2026-05-19T01:01:40 [2370519] | #25 0x000071385ee9ca94 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:447
# 2026-05-19T01:01:40 [2370519] | #26 0x000071385ef29a34 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```
RR trace is present on SDP:-
`/data/results/1779177603/MDEV-28730_07`

mariadb-SaahilAlam replied:
This is not reproducible on latest commit

mariadb-SaahilAlam commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,468 @@
+  else m_thr->lock_state= QUE_THR_LOCK_ROW;
+  if (m_trx->lock.wait_thr)
+  {
+    dberr_t wait_err= lock_wait(m_thr);
```
Server crash found here on debug build:-
`origin/main-MDEV-28730 ebbcbd200a4a62ad3dbfd132a83ef927a0cab648 `


```
Thread 2 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 3946709.3954078]
thd_storage_lock_wait (thd=0x0, value=12) at /data/Server/MDEV-28730C/sql/sql_class.cc:451
451       thd->utime_after_lock+= value;
(rr) set print addr off
(rr) bt
#0  thd_storage_lock_wait (thd=, value=12) at /data/Server/MDEV-28730C/sql/sql_class.cc:451
#1  lock_sys_t::wait_resume (this=this@entry=<lock_sys>, thd=, start=start@entry=..., now=...) at /data/Server/MDEV-28730C/storage/innobase/lock/lock0lock.cc:2108
#2  lock_wait (thr=<optimized out>) at /data/Server/MDEV-28730C/storage/innobase/lock/lock0lock.cc:2410
#3  QueryExecutor::handle_wait (this=this@entry=, err=DB_LOCK_WAIT, table_lock=table_lock@entry=false) at /data/Server/MDEV-28730C/storage/innobase/row/row0query.cc:189
#4  QueryExecutor::insert_record (this=this@entry=, table=table@entry=, tuple=tuple@entry=) at /data/Server/MDEV-28730C/storage/innobase/row/row0query.cc:155
#5  FTSQueryExecutor::insert_aux_record (this=this@entry=, aux_index=aux_index@entry=2 '\002', aux_data=aux_data@entry=) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0exec.cc:235
#6  fts_write_node (executor=executor@entry=, selected=selected@entry=2 '\002', aux_data=aux_data@entry=) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0fts.cc:3278
#7  fts_sync_write_words (executor=executor@entry=, index_cache=index_cache@entry=, unlock_cache=true) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0fts.cc:3348
#8  fts_sync_index (executor=executor@entry=, sync=sync@entry=, index_cache=index_cache@entry=) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0fts.cc:3399
#9  fts_sync (sync=, unlock_cache=<optimized out>, wait=wait@entry=true) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0fts.cc:3631
#10 fts_sync_table (table=, wait=wait@entry=true) at /data/Server/MDEV-28730C/storage/innobase/fts/fts0fts.cc:3704
#11 ha_innobase::optimize (this=, thd=) at /data/Server/MDEV-28730C/storage/innobase/handler/ha_innodb.cc:15326
#12 handler::ha_optimize (this=, thd=thd@entry=, check_opt=check_opt@entry=) at /data/Server/MDEV-28730C/sql/handler.cc:5884
#13 mysql_admin_table (thd=thd@entry=, tables=tables@entry=, check_opt=check_opt@entry=, operator_name=operator_name@entry=<msg_optimize>, lock_type=lock_type@entry=TL_WRITE, 
    org_open_for_modify=org_open_for_modify@entry=true, no_errors_from_open=false, extra_open_options=0, prepare_func=, operator_func=<optimized out>, view_operator_func=, is_cmd_replicated=true)
    at /data/Server/MDEV-28730C/sql/sql_admin.cc:949
#14 Sql_cmd_optimize_table::execute (this=<optimized out>, thd=) at /data/Server/MDEV-28730C/sql/sql_admin.cc:1730
#15 mysql_execute_command (thd=thd@entry=, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=false) at /data/Server/MDEV-28730C/sql/sql_parse.cc:5902
#16 mysql_parse (thd=thd@entry=, rawbuf=<optimized out>, length=<optimized out>, parser_state=parser_state@entry=) at /data/Server/MDEV-28730C/sql/sql_parse.cc:7942
#17 dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=, packet=packet@entry="", packet_length=packet_length@entry=83, blocking=blocking@entry=true) at /data/Server/MDEV-28730C/sql/sql_parse.cc:1898
#18 do_command (thd=thd@entry=, blocking=blocking@entry=true) at /data/Server/MDEV-28730C/sql/sql_parse.cc:1432
#19 do_handle_one_connection (connect=<optimized out>, connect@entry=, put_in_cache=put_in_cache@entry=true) at /data/Server/MDEV-28730C/sql/sql_connect.cc:1503
#20 handle_one_connection (arg=) at /data/Server/MDEV-28730C/sql/sql_connect.cc:1415
#21 start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:447
#22 clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```

RR trace is present on SDP:-
`/data/results/1779260294/MDEV-28730_08`

mariadb-SaahilAlam replied:
This is not reproducible on latest commit

mariadb-SaahilAlam commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11475,9 +11550,24 @@ innobase_fts_load_stopword(
+    if (success)
+      trx_commit_for_mysql(trx);
+    else trx->rollback();
+    trx->free();
```
Assertion found here:-
`MDEV-28730 CS 13.0.1 ebbcbd200a4a62ad3dbfd132a83ef927a0cab648 (Debug, Clang 18.1.3-11) `

```
SET @@GLOBAL.INNODB_flush_log_at_trx_commit=0;
CREATE TABLE t1 (c1 INT KEY,c2 VARCHAR(1),c3 TEXT,FULLTEXT KEY ftidx_9244(c2),FULLTEXT KEY ftidx_8094(c3)) ENGINE=INNODB DEFAULT CHARSET=utf8mb4;
```

`mariadbd: /test/main-MDEV-28730_dbg/storage/innobase/trx/trx0trx.cc:397: void trx_t::free(): Assertion `!commit_lsn' failed.
`

Thirunarayanan replied:
Fixed

iMineLink commented on storage/innobase/row/row0query.cc:
```diff
@@ -0,0 +1,577 @@
+      if (new_len != UNIV_SQL_NULL && new_len != old_len)
+      {
+        cmpl_info &= ~UPD_NODE_NO_SIZE_CHANGE;
+        err= DB_OVERFLOW;
```
I see here overflow is just a shortcut to go to optimistic update. Final results could be also underflow or other based on optimistic update logic.

iMineLink commented on storage/innobase/row/row0sel.cc:
```diff
@@ -4339,6 +4413,7 @@ It also has optimization such as pre-caching the rows, using AHI, etc.
 				pcur with stored position! In opening of a
 				cursor 'direction' should be 0.
 @return DB_SUCCESS or error code */
+template<typename Callback>
```
It seems, when `!Callback::needs_conversion`, that `buf` holds the `RecordCallback*` and not the MySQL row buffer. I think assertions should be added at the top of the function (gated by `if constexpr`) that check that the contract for `prebuilt` is respected, that is `!prebuilt->idx_cond` and `!prebuilt->pk_filter` (maybe also `!prebuilt->mysql_template`). This will prevent wrong usage I guess.

iMineLink commented on storage/innobase/include/fts0exec.h:
```diff
@@ -0,0 +1,420 @@
+class CommonTableReader : public RecordCallback
+{
+private:
+  std::vector<doc_id_t> doc_ids;
```
Noting this is not allocated on the InnoDB heap but using the global C++ allocator, and it seems pushback `doc_ids.push_back(mach_read_from_8(rec));` is done without checking for a max size before inserting

iMineLink commented on storage/innobase/fts/fts0exec.cc:
```diff
@@ -0,0 +1,941 @@
+  fts_write_doc_id(&last_doc_id, aux_data->last_doc_id);
+  dfield_set_data(field, &last_doc_id, sizeof(doc_id_t));
+
+  /* Field 5: doc_count (UINT32_T) */
```
maybe worth a static assertion that `sizeof(aux_data->doc_count) == 4`?

iMineLink commented on storage/innobase/fts/fts0exec.cc:
```diff
@@ -0,0 +1,941 @@
+
+  /* Field 3: value (CHAR(200)) */
+  field= dtuple_get_nth_field(&insert_tuple, 3);
+  dfield_set_data(field, value, strlen(value));
```
Unsure that `strlen()` is the right choice here, maybe `value->f_len` should be passed from the caller (same applies below).

Thirunarayanan replied:
update_config_record() callers pass the string doesn't contain NULL bytes. I think it is fine.

iMineLink commented on storage/innobase/fts/fts0exec.cc:
```diff
@@ -0,0 +1,941 @@
+
+  ulint doc_count_len;
+  const byte* doc_count_data= rec_get_nth_field(rec, offsets, 5, &doc_count_len);
+  ulint doc_count= mach_read_from_4(doc_count_data);
```
I wonder if this can be `uint32_t` instead.

iMineLink commented on storage/innobase/fts/fts0exec.cc:
```diff
@@ -0,0 +1,941 @@
+{
+  ut_ad(m_table);
+  ut_ad(common_table || m_index);
+  const size_t dbname_len= m_table->name.dblen() + 1;
```
Is this safe to do without holding `dict_sys.latch`?

Thirunarayanan replied:
Yes all callers of this function holds MDL on the user table.

iMineLink commented on storage/innobase/fts/fts0opt.cc:
```diff
@@ -1508,19 +1116,23 @@ fts_word_free(
 
-	mem_heap_free(heap);
+	if (heap) {
+		mem_heap_free(heap);
```
I suppose this shouldn't be the shared heap `AuxRecordReader::m_words_heap`, which is freed contextually in `i_s_fts_index_table_fill_one_index()`.

Thirunarayanan replied:
yes, that's why I added heap if condition

iMineLink commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11588,6 +11589,81 @@ ha_innobase::update_create_info(
+    cache->stopword_info.cached_stopword=
+      rbt_create_arg_cmp(
+        sizeof(fts_tokenizer_word_t), innobase_fts_text_cmp,
+        &my_charset_latin1);
```
Is this fixed charset choice correct?

Thirunarayanan replied:
Use the same code as before the patch:
```
fts_load_default_stopword(
/*======================*/
        fts_stopword_t*         stopword_info)  /*!< in: stopword info */
{
        fts_string_t            str;
        mem_heap_t*             heap;
        ib_alloc_t*             allocator;
        ib_rbt_t*               stop_words;

        allocator = stopword_info->heap;
        heap = static_cast<mem_heap_t*>(allocator->arg);

        if (!stopword_info->cached_stopword) {
                stopword_info->cached_stopword = rbt_create_arg_cmp(
                        sizeof(fts_tokenizer_word_t), innobase_fts_text_cmp,
                        &my_charset_latin1);
        }

```

iMineLink replied:
Noted, thanks. It smells strange but OK to keep legacy behavior I guess.

iMineLink reviewed (commented):
As this PR is rather large, I'm still not done with the review but wanted to post a few pending comments anyway before finishing the review.
In general, I see that reusing `row_search_mvcc()` is a really clever way to prevent most of the headaches around these changes, especially when secondary indices are involved. Since it seems that testing was already done on this revision, please feel free to discard minor comments such as those about `doc_count`, thank you.

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -1815,88 +1819,78 @@ CREATE TABLE $FTS_PREFIX_CONFIG
-	fts_table.suffix = "CONFIG";
-	fts_get_table_name(&fts_table, fts_name, true);
-	pars_info_bind_id(info, "config_table", fts_name);
+    if (i == 2)
```
Maybe add `ut_ad(!strcmp(fts_common_tables[2], "CONFIG"));` ?

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -2426,70 +2393,48 @@ fts_get_next_doc_id(
+  dberr_t error= executor->read_config_with_lock("synced_doc_id", reader);
+  if (error == DB_SUCCESS)
+  {
+    char value_buf[32];
```
Maybe this is `FTS_MAX_INT_LEN` ?

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -3742,157 +3309,85 @@ void fts_doc_ids_sort(ib_vector_t *doc_ids)
+  ut_a(n_elems > 0);
+  fts_doc_ids_sort(doc_ids);
+  dberr_t error= DB_SUCCESS;
+  for (uint32_t i= 0; i < n_elems && error == DB_SUCCESS; ++i)
```
Probably nothing changes, but `i` may be `ulint` instead to match `n_elems`.

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -3978,13 +3474,10 @@ fts_sync_index_reset(
-/** Commit the SYNC, change state of processed doc ids etc.
-@param[in,out]	sync	sync state
-@return DB_SUCCESS if all OK */
+/** Commit the SYNC, change state of processed doc ids etc.\n@param[in,out]\texecutor\t\tquery executor\n@param[in,out]\tsync\t\t\tsync state\n@return DB_SUCCESS if all OK */
```
I think this doxygen comment needs proper formatting

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -5584,288 +5024,267 @@ FTS.
+  dfield_t fields[1];
+  dtuple_t search_tuple{0, 1, 1, 0, fields, nullptr
+#ifdef UNIV_DEBUG
+                        , DATA_TUPLE_MAGIC_N
```
I wonder if it's worth it to introduce a macro that does this under the hood, as I see this pattern is repeated multiple times in the commit.

Thirunarayanan replied:
Is it worth it?

iMineLink replied:
Maybe not, OK to skip

iMineLink commented on storage/innobase/fts/fts0fts.cc:
```diff
@@ -5584,288 +5024,267 @@ FTS.
+    const byte* doc_id_data= rec_get_nth_field(rec, offsets, doc_col_pos, &len);
 
-	fts_doc_free(&doc);
+    if (len == sizeof(doc_id_t))
```
The contrary `len != sizeof(doc_id_t)` is possible only on corruption?

Thirunarayanan replied:
yes, record len could be wrong in case of corruption

iMineLink commented on storage/innobase/fts/fts0opt.cc:
```diff
@@ -779,241 +481,181 @@ fts_zip_deflate_end(
+          }
+          optim->trx->error_state = DB_SUCCESS;
+        }
+	else
```
Formatting is strange here, may be improved

iMineLink commented on storage/innobase/handler/i_s.cc:
```diff
@@ -2890,6 +2819,96 @@ i_s_fts_index_table_fill_one_fetch(
+          trx->error_state= DB_SUCCESS;
+          /* Restart transaction after rollback to restore will_lock flag */
+          trx_start_internal(trx);
+	  /* Clear words and retry */
```
I think the formatting is off here

iMineLink reviewed (commented):
I finished my review, all I have here are minor comments. I believe the only non-minor ones were the one (posted in previous review on the same commit) about default charset initialization to latin1 and table name construction without holding the `dict_sys` latch: please check those at least.
It's a great work, and reusing `row_search_mvcc()` is I think the best approach. Thanks!

iMineLink reviewed (approved):
Thanks for addressing my previous review points. LGTM now.

Thirunarayanan committed:
MDEV-28730 Remove internal parser usage from InnoDB FTS

InnoDB FTS performed all reads and DML on its auxiliary,
common and CONFIG tables through the InnoDB internal SQL
graph parser. Replace that path with direct B-tree access
via a new query-executor layer, and delete the parser-era
helpers.

row0query.h, row/row0query.cc - new QueryExecutor:
- General MVCC-aware record traversal and basic DML on the
clustered index. Sits on a btr_pcur and a transaction-owned
mtr; record processing goes through a RecordCallback that
bundles two std::functions:
  compare_record() returns SKIP / PROCESS / STOP for a record
  process_record() handles each PROCESS-ed (MVCC-visible) row

Public API:
  read()              scan a clustered index with a search key
  read_all()          full clustered scan (optional start tuple)
  read_by_index()     scan a secondary index, fetch the matching
                      clustered record, deliver it to the callback
  insert_record()     insert a tuple into the clustered index
  delete_record()     delete a row identified by tuple
  delete_all()        delete every row in the clustered index
  select_for_update() position+X-lock the matching clustered row
  update_record()     update the row select_for_update() locked,
                      falling back to optimistic/pessimistic and
                      external storage paths as needed
  replace_record()    upsert: select_for_update()+update_record(),
                      else insert_record()
  lock_table(), handle_wait(), commit_mtr()

fts0exec.h, fts/fts0exec.cc - new FTSQueryExecutor:
Thin wrapper over QueryExecutor specialised for FTS tables.
Opens and locks the required tables once and exposes typed
helpers keyed by table family.

Auxiliary INDEX_[1..6]:
  open_all_aux_tables()
  insert_aux_record(aux_index, fts_aux_data_t)
  delete_aux_record(aux_index, fts_aux_data_t)
  read_aux()         range scan from a given word
  read_from_range()  paginated read that absorbs
                     DB_FTS_EXCEED_RESULT_CACHE_LIMIT internally
                     and resumes from the last word seen

Common deletion tables (DELETED, DELETED_CACHE, BEING_DELETED,
                        BEING_DELETED_CACHE):
  open_all_deletion_tables()
  insert_common_record(), delete_common_record(),
  delete_all_common_records(), read_all_common()

CONFIG table (<key, value>):
  open_config_table() / set_config_table()
  insert_config_record(), update_config_record() (upsert),
  delete_config_record(), read_config_with_lock()

fts_aux_data_t carries the auxiliary row payload.
RecordCallback specialisations live alongside the executor:

CommonTableReader collects doc_ids from common tables that
share the <doc_id> schema.

ConfigReader extracts <key, value> and provides
compare_config_key() for fast key matching.

AuxRecordReader scans auxiliary indexes with an
AuxCompareMode (GREATER_EQUAL / GREATER / LIKE / EQUAL) driving the
comparator; tracks the last word seen so a paginated scan can resume.

fts_query() walks index and common tables via
QueryExecutor::read_by_index() with RecordCallback;

fts_write_node() writes auxiliary rows through
FTSQueryExecutor::insert_aux_record() / delete_aux_record()
with fts_aux_data_t.

fts_optimize_write_word() now goes through the same
insert/delete path.

fts_select_index{,_by_range,_by_hash} return uint8_t (was
ulint) with a simpler control flow.

fts_optimize_table() binds a thd to its transaction whether
invoked from a user thread or the FTS optimize thread.

fts_optimize_t drops its fts_index_table and fts_common_table
fts_table_t fields; fts_query_t drops fts_common_table.

storage/innobase/fts/fts0sql.cc is deleted along with the
commented-out and unreferenced parser-era helpers it held.

dict_sys.latch is now acquired once per fts_sync_table(),
fts_optimize_table() and fts_query() call to open every
auxiliary and common table in one pass, instead of being
re-acquired per table.

Every FTS trx_create() site now takes the THD from a real
caller (user trx, fts_opt_thd, DDL ctx->trx, ha_thd())
instead of either leaving it NULL or pulling current_thd.

fts_commit(): Each fts_commit_table() used to create transaction
and commit, producing N undo segments, N read views and N group-commit
batches for N FTS tables touched by the user statement.
fts_commit() now creates one internal trx, runs every
fts_commit_table() under it, and commits once.

fts_optimize_word() could fail when source nodes have equal
boundary doc_ids.  Loosen the merge predicate from '>' to
'>=' so legitimate overlapping ranges that occur when nodes
fill to FTS_ILIST_MAX_SIZE at doc_id boundaries are handled
correctly.

INFORMATION_SCHEMA.INNODB_FT_INDEX_TABLE (i_s.cc) wires the
shared words_heap via set_words_heap() and empties it per batch
with mem_heap_empty(); the per-fetch cleanup frees only the
per-node ilist (ut_alloc'd) and skips fts_word_free().
